README.md
ugrid2shp
===============================

This python code computes GIS shapefiles from a CF-UGRID compliant netCDF file.  The contours are generated in matplotlib, and output using Shapely and Fiona.

###System Requirements
- Python 2.7.x
- Shapely 
    - https://pypi.python.org/pypi/Shapely
- Fiona (Geospatial Data Abstraction Library (GDAL) OGR interface
    - https://pypi.python.org/pypi/Fiona
- libcurl 
    - OPeNDAP dependency. 
    - http://curl.haxx.se/libcurl/
- libhdf5 C library 
    - Dependency for NetCDF4.
    - http://www.hdfgroup.org/HDF5/release/obtain5.html
    - See notes section if installing on a mac - there is an easier way.
- netCDF4 C library 
    - http://www.unidata.ucar.edu/downloads/netcdf/index.jsp
    - See notes section if installing on a mac - there is an easier way.
- netcdf4-python 
    - The netCDF4/libhdf5 C libraries must be installed before installing netcdf4-python.
    - https://pypi.python.org/pypi/netCDF4
- shell-doctest 
    - This python library is needed to run doc tests.
    - https://pypi.python.org/pypi/shelldoctest

###Notes: 
On a mac using the homebrew package manager (http://brew.sh/), the homebrew/science tap (https://github.com/Homebrew/homebrew-science) can be used to simplify installation of the HDF5 and netCDF4 C libraries needed for building netcdf4-python. 
After installing homebrew, add the homebrew-science tap, then install the HDF5 and netCDF4 C libraries from the homebrew-science tap

###Usage:

ugrid2shp.py -h -d -s -w -x -z -n \<NcFileName|url\> -o \<OutFile\> -v \<NcVariableName\> -a \<MinVal\> -b \<MaxVal\> -c \<NumLevels\> -l \<AxisLims\> -p \<ProjectionString\>
		
	where:	-h | --Help				the text you are looking at right now
			-d | --Debug				display debugging diagnostics [False]
			-s | --Silent				no screen diagnostic output [True]
			-w | --WriteImage			write matplotlib image to a file <outfile>.png [False]
			-x | --ShowImage			display matplotlib image, user must close before continuing [False]
			-z | --NoZip				no zip file output [False]
			-n | --NcFileName	<ncfilename> 	netCDF file to read from, or a URL to an OPeNDAP file [maxele.63.nc]
			-o | --OutFile		<outfilename> 	filename to write shapefile to [outShape]
			-v | --NcVarName	<NcVarName> 	netCDF variable name to render [zeta_max]
			-a | --MinVal		<min value> 	smallest scalar value to render [0]
			-b | --MaxVal		<max value> 	largest scalar value to render [10]
			-c | --NumLevels	<num levels> 	number of contour levels [11]
			-l | --AxisLims		<[x0 y0 x1 y1]> axis limits to clip to [full domain]
			-p | --ProjStr				Projection string for prj file;  see below.
 
	A projection file (prj) is written containing the following:
		GEOGCS["GCS_WGS_1984",DATUM["D_WGS_1984",SPHEROID["WGS_1984",6378137.0,298.257223563]],PRIMEM["Greenwich",0.0],UNIT["Degree",0.0174532925199433]]
	To specify a different projection string, pass in -p | --ProjStr <projstr>, where projstr is as in this example:
		--ProjStr 'GEOGCS["GCS_WGS_1984",DATUM["D_WGS_1984",SPHEROID["WGS_1984",6378137.0,298.257223563]],PRIMEM["Greenwich",0.0],UNIT["Degree",0.0174532925199433]]' 

###Example usage:

    $ Define url=http://opendap.renci.org:1935/thredds/dodsC/ASGS/nam/2013112112/ec95d/hatteras.renci.org/cfsamp/namforecast/maxele.63.nc 
    $ Run silently (-s) pointing to the url devined above (-n $url), and with the maximum scalar value set to 6 (-b 6) 
    $ python ugrid2shp.py -s -n $url -b 6
    $ The first image is the matplotlib figure.  The second is the shapefile in QGIS.  the colormaps are not the same, but the contour lines are in the same locations.
![](test1.png)
![](QGIS_test1.png)


###Test URLs:
	Coarse resolution ADCIRC grid: 
 		http://opendap.renci.org:1935/thredds/dodsC/ASGS/nam/2013112112/ec95d/hatteras.renci.org/cfsamp/namforecast/maxele.63.nc
 	Medium resolution ADCIRC grid: 
 		http://opendap.renci.org:1935/thredds/dodsC/ASGS/andrea/08/nc6b/blueridge.renci.org/fivemem/nhcConsensus/maxele.63.nc





HISTORY.txt
Version 1.0          2015-04-23
-------------------------------
- Initial release. (cbc)


tests/__init__.py
"""The pyugrid.test subpackage."""

adcirc_local_test.py
#!/usr/local/bin/python
"""
$ python ugrid2shp.py -o test_local -a 0 -b 11
file://maxele.63.nc
Getting data from maxele.63.nc... 
Triangulating ...
Making contours in figure ...
Calling tricontourf  ...
Extracting contour shapes from tricontourf object ...
Writing shapes to test_local.shp
Writing prj to test_local.prj
Flushing zip file to test_local.zip
"""
if __name__ == '__main__':
	import shelldoctest
	shelldoctest.testmod(verbose=3)


adcirc_thredds_test.py
"""
$ python ugrid2shp.py -n http://opendap.renci.org:1935/thredds/dodsC/ASGS/nam/2014112112/nc6b/hatteras.renci.org/rencidaily/namforecast/maxele.63.nc  -o test_thredds -a 0 -b 11 
http://opendap.renci.org:1935/thredds/dodsC/ASGS/nam/2014112112/nc6b/hatteras.renci.org/rencidaily/namforecastmaxele.63.nc
Getting data from http://opendap.renci.org:1935/thredds/dodsC/ASGS/nam/2014112112/nc6b/hatteras.renci.org/rencidaily/namforecast/maxele.63.nc...
Triangulating ...
Making contours in figure ...
Calling tricontourf  ...
Extracting contour shapes from tricontourf object ...
Writing shapes to test_thredds.shp
Writing prj to test_thredds.prj
Flushing zip file to test_thredds.zip
"""
if __name__ == '__main__':
	import shelldoctest
	shelldoctest.testmod(verbose=3)


netcdf_viz.py
from pylab import *
import matplotlib.tri as Tri
from shapely.geometry import mapping, Polygon
import fiona
import netCDF4
import datetime
import time
import sys
import getopt

def main(argv):
    
    # Defaults parameter values.
    ncfilename=None
    outfile='outShape'
    NcVariableName='zeta_max'
    MinVal = 0
    MaxVal = 11
    AxisLims = []
    NumberOfSamples = int(MaxVal - MinVal)
    Tflag = False

    try:
        opts, args = getopt.getopt(argv,"hn:o:v:a:b:c:s:t:",["ncfilename=","outfile=", "NcVariableName=", "MinVal=", "MaxVal=", "NumberOfSamples=", "AxisLims=", "TimeStepNum"])
    except getopt.GetoptError:
        print 'Option error: '
        print 'Run adcirc_netcdf_viz.py -h for help'
        sys.exit(2)
    for opt, arg in opts:
        if opt == '-h':
            print 'adcirc_netcdf_viz.py -n <ncfilename> -o <outfile> -v <NcVariableName> -a <MinVal> -b <MaxVal> -c <NumberOfSamples> -s <AxisLims> -t TimeStepNum'
            sys.exit()
        elif opt in ("-n", "--ncfilename"):
            ncfilename = arg
        elif opt in ("-o", "--outfile"):
            outfile = arg
        elif opt in ("-v", "--NcVariableName"):
            NcVariableName = arg
        elif opt in ("-a", "--MinVal"):
            MinVal = arg
        elif opt in ("-b", "--MaxVal"):
            MaxVal = arg
        elif opt in ("-c", "--NumberOfSamples"):
            NumberOfSamples = int(arg)
        elif opt in ("-s", "--AxisLims"):
            AxisLims = arg
        elif opt in ("-t", "--TimeStepNum"):
            Tflag = True
            TimeStepNum = int(arg)

    MinVal = float(MinVal)
    MaxVal = float(MaxVal)
    imagefilename=outfile+'.png'
    shapefilename=outfile+'.shp'
    prjfilename=outfile+'.prj'

    if ncfilename == None:
        raise Exception("You must pass a netCDF4 filename as an argument. Use the -n or --ncfilename argument to do this.")
    url = ncfilename
    titl='ADCIRC';
    if url[-3:] != '.nc':
        url=url+'.nc'
    vname=NcVariableName
    

    print 'Getting data from url=%s... ' % url

    nc=netCDF4.Dataset(url).variables
    nc.keys()
    print nc.keys()
    lon = nc['x'][:]
    lat = nc['y'][:]
    # Move to 0-indexing by subtracting 1
    nv = nc['element'][:,:]-1

    print "Shape of lon is (%i)" % lon.shape
    print "Shape of lat is (%i)" % lat.shape
    print "Shape of nv is (%i, %i)" % nv.shape

    print "Max of lon is (%i)" % lon.max()
    print "Max of lat is (%i)" % lat.max()
    print "Max of nv is (%i)" % nv.max()

    var = nc[vname][:]

    if Tflag == True:
        start=datetime.datetime(2008, 9, 13, 6, 0, 0)
        time_var = nc['time']
        print "The time chosen is: "
        print time_var[TimeStepNum]
        print "Units are: "
        print time_var.units
        if "since" in time_var.units:
            since_when = time_var.units.rsplit(None, 1)[-1]
            since_when_string = eval("time_var."+since_when)
            print since_when + " is " + since_when_string
        var = nc[vname][TimeStepNum,:]
    
    print "var[0]: " + str(var[0])
    print 'Triangulating ...'
    tri = Tri.Triangulation(lon,lat, triangles=nv)

    print 'Making contours in figure ...'
    figure(figsize=(10,10))
    subplot(111,aspect=(1.0/cos(mean(lat)*pi/180.0)))
    levels = linspace(MinVal, MaxVal, num=NumberOfSamples)
    print 'Calling tricontourf  ...'
    contour = tricontourf(tri, var,levels=levels,shading='faceted')
    colorbar(orientation='horizontal')
    title(url)
    # This takes the axis limit string arg and converts it to a list. Then it converts that list to floats.
    if AxisLims != []:
        AxisLimsSplit = AxisLims.split(', ')
        AxisLims = map(float, AxisLimsSplit)
        axis(AxisLims)

    # Uncomment to show matplotlib visualization.
    # show()

    print "Saving figure as " +  imagefilename
    savefig(imagefilename)

    print 'Extracting contour shapes from tricontourf object ...'
    geoms = []
    # create list of tuples (geom, vmin, vmax)
    for colli,coll in enumerate(contour.collections):
        vmin,vmax = contour.levels[colli:colli+2]
        for p in coll.get_paths():
            p.simplify_threshold = 0.0
            polys = p.to_polygons()
            # Don't append an empty polygon to the geometry or it will screw up the indexing.
            if polys:
                geoms.append( (Polygon(polys[0],polys[1:] ),vmin,vmax))

    print "Writing shapes to " + shapefilename
    schema = { 'geometry': 'Polygon', 'properties': { 'vmin': 'float', 'vmax': 'float' } }
    with fiona.open(shapefilename, 'w', 'ESRI Shapefile', schema) as c:
        for geom in geoms:
            c.write({
                'geometry': mapping(geom[0]),
                'properties': {'vmin': geom[1], 'vmax': geom[2]},
            })

    print "Writing prj to " + prjfilename
    prj_string = '''GEOGCS["GCS_WGS_1984",DATUM["D_WGS_1984",SPHEROID["WGS_1984",6378137.0,298.257223563]],PRIMEM["Greenwich",0.0],UNIT["Degree",0.0174532925199433]]'''
    prj_file = open(prjfilename, 'w')
    prj_file.write(prj_string)
    prj_file.close()

    # try reading the Shapefile back in 
    c = fiona.open(shapefilename, 'r')

if __name__ == "__main__":
    main(sys.argv[1:])


adcirc_netcdf_viz.py
from pylab import *
import matplotlib.tri as Tri
from shapely.geometry import mapping, Polygon
import fiona
import netCDF4
import datetime
import time
import sys
import getopt

def main(argv):
    
    # Defaults parameter values.
    ncfilename='temp'
    outfile='outShape'
    NcVariableName='zeta_max'
    MinVal = 0
    MaxVal = 11
    AxisLims = []

    try:
        opts, args = getopt.getopt(argv,"hn:o:v:a:b:s:",["ncfilename=","outfile=", "NcVariableName=", "MinVal=", "MaxVal=", "AxisLims="])
    except getopt.GetoptError:
        print 'Option error: '
        print 'adcirc_netcdf_viz.py -n <ncfilename> -o <outfile> -v <NcVariableName> -a <MinVal> -b <MaxVal> -s <AxisLims>'
        sys.exit(2)
    for opt, arg in opts:
        if opt == '-h':
            print 'adcirc_netcdf_viz.py -n <ncfilename> -o <outfile> -v <NcVariableName> -a <MinVal> -b <MaxVal> -s <AxisLims>'
            sys.exit()
        elif opt in ("-n", "--ncfilename"):
            ncfilename = arg
        elif opt in ("-o", "--outfile"):
            outfile = arg
        elif opt in ("-v", "--NcVariableName"):
            NcVariableName = arg
        elif opt in ("-a", "--MinVal"):
            MinVal = arg
        elif opt in ("-b", "--MaxVal"):
            MaxVal = arg
        elif opt in ("-s", "--AxisLims"):
            AxisLims = arg

    MinVal = float(MinVal)
    MaxVal = float(MaxVal)
    imagefilename=outfile+'.png'
    shapefilename=outfile+'.shp'
    prjfilename=outfile+'.prj'

    titl='ADCIRC';
    url=ncfilename+'.nc'
    vname=NcVariableName
    start=datetime.datetime(2008, 9, 13, 6, 0, 0)

    print 'Getting data from url=%s... ' % url

    time0=time.time()
    nc=netCDF4.Dataset(url).variables
    nc.keys()
    print nc.keys()
    lon = nc['x'][:]
    lat = nc['y'][:]
    # Move to 0-indexing by subtracting 1
    nv = nc['element'][:,:]-1

    print "Shape of lon is (%i)" % lon.shape
    print "Shape of lat is (%i)" % lat.shape
    print "Shape of nv is (%i, %i)" % nv.shape

    print "Max of lon is (%i)" % lon.max()
    print "Max of lat is (%i)" % lat.max()
    print "Max of nv is (%i)" % nv.max()

    # time_var = nc['time']
    # dtime = netCDF4.num2date(time_var[:],time_var.units)
    #istart = netCDF4.date2index(start,time_var,select='nearest')
    #istart=1
    #var = nc[vname][istart,:]    
    var = nc[vname][:]
    print 'elapsed time= %d seconds' % (time.time()-time0)
    print "var[0]: " + str(var[0])

    print 'Triangulating ...'
    time0=time.time()
    tri = Tri.Triangulation(lon,lat, triangles=nv)
    print 'elapsed time= %d seconds' % (time.time()-time0)

    print 'Making contours in figure ...'
    time0=time.time()
    figure(figsize=(10,10))
    subplot(111,aspect=(1.0/cos(mean(lat)*pi/180.0)))
    levels = linspace(MinVal, MaxVal)
    print 'Calling tricontourf  ...'
    contour = tricontourf(tri, var,levels=levels,shading='faceted')
    colorbar(orientation='horizontal')
    title(url)
    print 'elapsed time= %d seconds' % (time.time()-time0)
    # This takes the axis limit string arg and converts it to a list. Then it converts that list to floats.
    if AxisLims != []:
        AxisLimsSplit = AxisLims.split(', ')
        AxisLims = map(float, AxisLimsSplit)
        axis(AxisLims)

    # Uncomment to show matplotlib visualization.
    # show()

    print "Saving figure as " +  imagefilename
    savefig(imagefilename)

    time0=time.time()
    print 'Extracting contour shapes from tricontourf object ...'
    geoms = []
    # create list of tuples (geom, vmin, vmax)
    for colli,coll in enumerate(contour.collections):
        vmin,vmax = contour.levels[colli:colli+2]
        for p in coll.get_paths():
            p.simplify_threshold = 0.0
            polys = p.to_polygons()
            # Don't append an empty polygon to the geometry or it will screw up the indexing.
            if polys:
                geoms.append( (Polygon(polys[0],polys[1:] ),vmin,vmax))

    print "Writing shapes to " + shapefilename
    schema = { 'geometry': 'Polygon', 'properties': { 'vmin': 'float', 'vmax': 'float' } }
    with fiona.open(shapefilename, 'w', 'ESRI Shapefile', schema) as c:
        for geom in geoms:
            c.write({
                'geometry': mapping(geom[0]),
                'properties': {'vmin': geom[1], 'vmax': geom[2]},
            })

    print 'elapsed time= %d seconds' % (time.time()-time0)
    print "Writing prj to " + prjfilename
    prj_string = '''GEOGCS["GCS_WGS_1984",DATUM["D_WGS_1984",SPHEROID["WGS_1984",6378137.0,298.257223563]],PRIMEM["Greenwich",0.0],UNIT["Degree",0.0174532925199433]]'''
    prj_file = open(prjfilename, 'w')
    prj_file.write(prj_string)
    prj_file.close()

    # try reading the Shapefile back in 
    c = fiona.open(shapefilename, 'r')

if __name__ == "__main__":
    main(sys.argv[1:])


ugrid2shp/ugrid2shp.py
#
# ugrid2shp.py -  shapefile generator for CF-UGRID compliant netCDF files
#
# Idea from Rusty Holleman and Rich Signell (shapely/fiona)
# Code by Brian Blanton, RENCI
#

from pylab import *
import matplotlib.tri as Tri
import matplotlib.pyplot as plt
from shapely.geometry import mapping, Polygon
import fiona
import netCDF4
import datetime
import time
import sys
import argparse
import zipfile
import os

defaults = {'ProjStr': 'GEOGCS["GCS_WGS_1984",'
                       'DATUM["D_WGS_1984",'
                       'SPHEROID["WGS_1984",6378137.0,298.257223563]],'
                       'PRIMEM["Greenwich",0.0],'
                       'UNIT["Degree",0.0174532925199433]]',
            'axis_limits': [],
            'debug': False,
            'maxval': 10,
            'minval': 0,
            'nc_var_name': 'zeta_max',
            'ncfilename': 'maxele.63.nc',
            'no_zip': False,
            'numlevels': 11,
            'outfilename': 'outShape',
            'show_image': False,
            'silent': False,
            'write_image': False}


def parse_args(argv):
    parser = argparse.ArgumentParser(
        formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    parser.add_argument('-d', '--debug',
                        dest='debug',
                        action='store_true',
                        help='Display diagnostics.')
    parser.add_argument('-s', '--silent',
                        dest='silent',
                        action='store_true',
                        help='No screen diagnostic output.')
    parser.add_argument('-w', '--writeimage',
                        dest='write_image',
                        action='store_true',
                        help='Write matplotlib image to a file <outfile>.png.')
    parser.add_argument('-x', '--showimage',
                        dest='show_image',
                        action='store_true',
                        help='Display matplotlib image, '
                             'user must close before continuing.')
    parser.add_argument('-z', '--nozip',
                        dest='no_zip',
                        action='store_true',
                        help='No zip file output.')
    parser.add_argument('-n', '--ncfilename',
                        dest='ncfilename',
                        action='store',
                        metavar='NC_FILE_NAME',
                        default=defaults['ncfilename'],
                        help='Path to NetCDF file to read, '
                             'or a URL to an OPeNDAP resource.')
    parser.add_argument('-o', '--outfilename',
                        dest='outfilename',
                        action='store',
                        metavar='OUTPUT_FILE_NAME',
                        default=defaults['outfilename'],
                        help='Filename to write as shapefile.')
    parser.add_argument('-v', '--ncvarname',
                        dest='nc_var_name',
                        action='store',
                        metavar='NC_VAR_NAME',
                        default=defaults['nc_var_name'],
                        help='NetCDF variable name to render.')
    parser.add_argument('-a', '--minval',
                        type=int,
                        dest='minval',
                        action='store',
                        metavar="MIN_VAL",
                        default=defaults['minval'],
                        help='Smallest scalar value to render.')
    parser.add_argument('-b', '--maxval',
                        type=int,
                        dest='maxval',
                        action='store',
                        metavar='MAX_VAL',
                        default=defaults['maxval'],
                        help='Largest scalar value to render.')
    parser.add_argument('-c', '--numlevels',
                        type=int,
                        dest='numlevels',
                        action='store',
                        metavar='NUM_LEVELS',
                        default=defaults['numlevels'],
                        help='Number of contour levels.')
    parser.add_argument('-l', '--axislimits',
                        nargs=4,
                        dest='axis_limits',
                        action='store',
                        metavar=('X0', 'Y0', 'X1', "Y1"),
                        default=defaults['axis_limits'],
                        help='Bounding box list of coordinates.')
    parser.add_argument('-p', '--proj',
                        dest="ProjStr",
                        action='store',
                        metavar="PROJECTION_STRING",
                        default=defaults['ProjStr'],
                        help="Projection string as Well Known Text (WKT).")

    return parser.parse_args(argv)


def ugrid2shp(argv):

    args = parse_args(argv)

    imagefilename = args.outfilename + '.png'
    shapefilename = args.outfilename + '.shp'
    prjfilename = args.outfilename + '.prj'
    zipfilename = args.outfilename + '.zip'
    readmefilename = 'README.txt'

    # get timestamp
    ts = time.time()
    st = datetime.datetime.fromtimestamp(ts).strftime('%Y-%m-%d %H:%M:%S')
    if args.debug:
        print "Create time " + st

    url = args.ncfilename
    path, ncfile = os.path.split(url)
    if not path:
        path = 'file://'
    if not args.silent:
        print path + ncfile

    # end of url must have .nc
    # if url[-3:] != '.nc': url=url+'.nc'

    # get stuff from netCDF file
    if not args.silent:
        print 'Getting data from %s... ' % url

    nc = netCDF4.Dataset(url)
    ncvars = nc.variables
    if args.debug:
        print ncvars.keys()

    # get some global attributes
    model = getattr(nc, 'model')
    grid = getattr(nc, 'agrid')
    inst = getattr(nc, 'institution')

    lon = ncvars['x'][:]
    lat = ncvars['y'][:]
    var = ncvars[args.nc_var_name]
    units = var.units
    data = ncvars[args.nc_var_name][:]
    if len(data.shape) > 1:
        data = data[0, :]
    elems = ncvars['element'][:, :] - 1  # Move to 0-indexing by subtracting 1
    plot_title = 'MatPlotLib plot of ' + args.nc_var_name + ' in ' + ncfile

    if args.debug:
        print "\n   Shape of lon is (%i)" % lon.shape
        print "   Shape of lat is (%i)" % lat.shape
        print "   Shape of nv is (%i, %i)" % elems.shape
        print "   Max of lon is (%i)" % lon.max()
        print "   Max of lat is (%i)" % lat.max()
        print "   Max of nv is (%i)" % elems.max()
        print "   First value in var: " + str(data[0]) + '\n'

    if not args.silent:
        print 'Triangulating ...'
    tri = Tri.Triangulation(lon, lat, triangles=elems)

    if not args.silent:
        print 'Making contours in figure ...'
    fig = plt.figure(figsize=(10, 10))
    plt.subplot(111, aspect=(1.0 / cos(mean(lat) * pi / 180.0)))
    levels = linspace(args.minval, args.maxval, num=args.numlevels)

    if not args.silent:
        print 'Calling tricontourf  ...'
    contour = tricontourf(tri, data, levels=levels, shading='faceted')
    cbar = fig.colorbar(contour, orientation='vertical', ticks=levels)
    cbar.set_label(units)
    title(plot_title)
    figtext(.5,
            .05,
            'URL: ' + path + ncfile,
            size='small',
            horizontalalignment='center')

    # This takes the axis limit string arg and converts it to a list.
    # Then it converts that list to floats.
    if args.axis_limits:
        axis_limits = map(float, args.axis_limits)
        axis(axis_limits)

    if args.write_image:
        if not args.silent:
            print "Saving figure as " + imagefilename
        savefig(imagefilename)

    if args.show_image:
        print "\nDisplay window must be closed " \
              "before shapefile can be written. \n"
        show()

    # this is the meat, as per Rusty Holleman
    if not args.silent:
        print 'Extracting contour shapes from tricontourf object ...'
    geoms = []
    # create list of tuples (geom, vmin, vmax)
    for colli, coll in enumerate(contour.collections):
        vmin, vmax = contour.levels[colli:colli + 2]
        for p in coll.get_paths():
            p.simplify_threshold = 0.0
            polys = p.to_polygons()
            # Don't append an empty polygon to the geometry
            # or it will screw up the indexing.
            if polys:
                # Don't append holes without at least three coordinates
                polys[1:] = [poly for poly in polys[1:] if poly.shape[0] > 2]
                geoms.append((Polygon(polys[0], polys[1:]), vmin, vmax))

    # this is the other meat, as per Rusty Holleman
    if not args.silent:
        print "Writing shapes to " + shapefilename
    schema = {'geometry': 'Polygon',
              'properties': {'vmin': 'float',
                             'vmax': 'float'}}
    with fiona.open(shapefilename, 'w', 'ESRI Shapefile', schema) as c:
        for geom in geoms:
            c.write({
                'geometry': mapping(geom[0]),
                'properties': {'vmin': geom[1], 'vmax': geom[2]},
            })

    if not args.silent:
        print "Writing prj to " + prjfilename

    prj_file = open(prjfilename, 'w')
    prj_file.write(args.ProjStr)
    prj_file.close()

    # write a README file
    rmf = open(readmefilename, mode='w')
    rmf.write('CreateTime: {0}\n'.format(st))
    rmf.write('Model: {0}\n'.format(model))
    rmf.write('Grid: {0}\n'.format(grid))
    rmf.write('Institution: {0}\n'.format(inst))
    rmf.write('Url: {0}\n'.format(path))
    rmf.write('File: {0}\n'.format(ncfile))
    rmf.write('ProjStr: {0}\n'.format(args.ProjStr))
    rmf.write('Units: {0}\n'.format(units))
    rmf.close()

    if not args.no_zip:
        if not args.silent:
            print 'Flushing zip file to ' + zipfilename
        zf = zipfile.ZipFile(zipfilename, mode='w')
        zf.write(args.outfilename + '.shp')
        if os.path.isfile(args.outfilename + '.prj'):
            zf.write(args.outfilename + '.prj')
        if os.path.isfile(args.outfilename + '.cpg'):
            zf.write(args.outfilename + '.cpg')
        if os.path.isfile(args.outfilename + '.shx'):
            zf.write(args.outfilename + '.shx')
        if os.path.isfile(args.outfilename + '.dbf'):
            zf.write(args.outfilename + '.dbf')
        zf.write(readmefilename)
        zf.close()

        # Remove working files
        if os.path.isfile(args.outfilename + '.prj'):
            os.remove(args.outfilename + '.prj')
        if os.path.isfile(args.outfilename + '.cpg'):
            os.remove(args.outfilename + '.cpg')
        if os.path.isfile(args.outfilename + '.shx'):
            os.remove(args.outfilename + '.shx')
        if os.path.isfile(args.outfilename + '.dbf'):
            os.remove(args.outfilename + '.dbf')
        os.remove(args.outfilename + '.shp')

    # try reading the Shapefile back in
    # c = fiona.open(shapefilename, 'r')
